### PR TITLE
Refactor RNTester

### PR DESCRIPTION
**Pull request by anku255**
## Summary

This PR tries to remove the code duplication between `RNTesterApp.ios.js` and `RNTesterApp.android.js`. The common code has been moved to `RNTesterAppShared.js`. 

This PR also refactors the state architecture and simplifies it to use just one single `useReducer` hook to manage the app's state.

It also adds the visibility hack to make the screen change a little bit faster and removed the unneeded test files.

### CODE REVIEW

File: packages/rn-tester/js/utils/index.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
+/**
```
Comment (rickhanlonii) [inline packages/rn-tester/js/utils/index.js:1]: Can we rename this file, with Haste every file name needs to be unique (across all of facebook). Something like `testerStateUtils`

File: packages/rn-tester/js/RNTesterAppShared.js [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+import {RNTesterEmptyBookmarksState} from './components/RNTesterEmptyBookmarksState';
+
+import type {RNTesterTheme} from './components/RNTesterTheme';
+import type {ExamplesList} from './types/RNTesterTypes';
+
+const APP_STATE_KEY = 'RNTesterAppState.v3';
+
+// RNTester App currently uses Async Storage from react-native for storing navigation state
+// and bookmark items.
+// TODO: Add Native Async Storage Module in RNTester
```
Comment (rickhanlonii) [inline packages/rn-tester/js/RNTesterAppShared.js:45]:
```suggestion
// TODO: Vendor AsyncStorage or create our own.
```

File: packages/rn-tester/js/RNTesterAppShared.js [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+LogBox.ignoreLogs([new RegExp('has been extracted from react-native')]);
```
Comment (rickhanlonii) [inline packages/rn-tester/js/RNTesterAppShared.js:46]: Can we either add AsyncStorage to this or list all of the libs that throw this individually? Also there's no need for the RegExp constructor, we could just do `/has been extracted from react-native/`.

File: packages/rn-tester/js/RNTesterAppShared.js [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+      initialStateFromStorage => {
+        dispatch({
+          type: RNTesterActionsType.INIT_FROM_STORAGE,
+          data: initialStateFromStorage,
+        });
+      },
+    );
+  }, [dispatch]);
+
+  const {openExample, screen, bookmarks, recentlyUsed} = state;
```
Comment (rickhanlonii) [inline packages/rn-tester/js/RNTesterAppShared.js:125]: Typically would see this at the top of the function with the rest of the state.

File: packages/rn-tester/js/RNTesterAppShared.js [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+    args => {
+      dispatch({
+        type: RNTesterActionsType.NAVBAR_PRESS,
+        data: {screen: args.screen},
+      });
+    },
+    [dispatch],
+  );
+
+  const colorScheme = useColorScheme();
```
Comment (rickhanlonii) [inline packages/rn-tester/js/RNTesterAppShared.js:180]: Same here, would move all the hooks you use up to the top of the function.

File: packages/rn-tester/js/components/RNTesterHeader.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+const HeaderIOS = ({
+  onBack,
+  title,
+  documentationURL,
+}: {
+  onBack?: () => mixed,
+  title: string,
+  documentationURL?: string,
+  ...
```
Comment (rickhanlonii) [inline packages/rn-tester/js/components/RNTesterHeader.js:33]: Is this not sealed? Could we remove the `...`?

File: packages/rn-tester/js/components/RNTesterNavbar.js [hunk 7]
```diff
@@ -5,31 +5,37 @@
  * LICENSE file in the root directory of this source tree.
  *
  * @format
+ * @flow
```
Comment (rickhanlonii) [inline packages/rn-tester/js/components/RNTesterNavbar.js:8]: Nice, thanks!

File: packages/rn-tester/js/RNTesterAppShared.js (hunk 3 above)
Comment (anku255) [reply packages/rn-tester/js/RNTesterAppShared.js]: Done!

File: packages/rn-tester/js/RNTesterAppShared.js (hunk 4 above)
Comment (anku255) [reply packages/rn-tester/js/RNTesterAppShared.js]: Done!

File: packages/rn-tester/js/RNTesterAppShared.js (hunk 5 above)
Comment (anku255) [reply packages/rn-tester/js/RNTesterAppShared.js]: Done!

File: packages/rn-tester/js/components/RNTesterHeader.js (hunk 6 above)
Comment (anku255) [reply packages/rn-tester/js/components/RNTesterHeader.js]: Done!

**Comment by anku255** [comment]
> LGTM!
> 
> I left some comments, we will need to at least renamed the index.js file to unblock the land.

@rickhanlonii Thank you for the reviewing. I have done the changes suggested by you. :)

File: packages/rn-tester/js/components/RNTesterEmptyBookmarksState.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
+import React from 'react';
```
Comment (rickhanlonii) [inline packages/rn-tester/js/components/RNTesterEmptyBookmarksState.js:1]: Can you add the copyright header to this file?

Comment (anku255) [reply packages/rn-tester/js/components/RNTesterEmptyBookmarksState.js]: Done!

**Pull request closed without merging**